### DISCUSSION TITLE
How to debug a circular dependency

### DISCUSSION THREAD

**Discussion by trampster** [YairHalberstadt/stronginject#125] [Q&A]
I'm getting an error:
Error while resolving dependencies for 'Pttox.Core.Services.ILinkHandler: 'Pttox.COre.Services.IPttClient' has a circular dependency.

Original I was trying to find a loop including ILinkHandler and IPttClient, but LinkHandler is only used by one thing and there is no loop. So I'm guess that ILinkHandler isn't part of the loop and that it is mentioned in the error only because it has a dependency on IPttClient.

Unfortunately IPttClient has a lot of dependencies and finding the loop is very hard (also our old IoC worked, without complaining about a loop)

Is there a way to get StringInject to print the dependency chain for the loop, so that I can debug this issue?

**Comment by trampster** [comment]
I've discovered a possible cause:

IPttClient -> Func<CallInfo, SessionInfo> -> IPttService  -> IPttClient

This wasn't a problem for our existing IoC because of the following:
* IPttClient is a singleton
* Func<CallInfo, SessionInfo> is lazy so it's dependencies don't need to be resolved until it is called, at which point IPttClient is already created.

Is it possible that this is what is causing my issue, If so how would you work around it? Is it possible to delay the resolution of IPttService until the func is called?

Func<CallInfo, SessionInfo> is defined as a Factory

```
[Factory(Scope.InstancePerResolution)]
public static Func<CallInfo, ISession> CreateSessionFactory(IPttService pttService)
{
    return callInfo => new Session(pttService);
}
```

**Reply by trampster** [reply]
It's definitely caused by this. Verified by removing IPttService from the CreateSessionFactory and making the func return null. This caused the error to go away.

I tried using Lazy on IPttService (Lazy<IPttService>) but this didn't help, the error still occurs and it also complains in a warning that it's a singleton.

**Reply by trampster** [reply]
I have found one workaround and that is to move the Factory definition out of the Module and into the container, then drop IPttService from the factory arguments and resolve it using this.Resolve<IPttService>()

This works but isn't ideal because it must now live in the Container which must be defined in the plaform specific project which means it must be duplicated for Android/iOS etc.

**Reply by YairHalberstadt** [reply]
Interesting point. If you open an issue for ignoring circular dependencies in this specific case I'll look into it when I have a chance.

**Reply by YairHalberstadt** [reply]
FIxed in #128
